### ISSUE TITLE
The new completer does not sort items on filtering

### ISSUE THREAD

**Issue by krassowski** [jupyterlab/jupyterlab#9048]
## Description

The new completer does not sort items. The old one did, see https://github.com/krassowski/jupyterlab-lsp/issues/369#issuecomment-695786331. This affects jupyterlab-lsp and jupyterlab-kite I think. Also, we should probably remove the old completer in v3 once this is resolved.

## Reproduce

Install jupyterlab-lsp, see:

![Screenshot from 2020-09-18 18-21-53](https://user-images.githubusercontent.com/5832902/93626868-f910c000-f9db-11ea-972a-c0aa7ad37923.png)

![Screenshot from 2020-09-18 18-24-12](https://user-images.githubusercontent.com/5832902/93626989-2fe6d600-f9dc-11ea-96ac-85e2d429a631.png)

![Screenshot from 2020-09-18 18-27-07](https://user-images.githubusercontent.com/5832902/93627271-9966e480-f9dc-11ea-8643-d893f7995fa7.png)

## Expected behavior

Transpose should be first in numpy example, geom_count should be first in R example.
`keys()` should be before `fromkeys()`

## Context

Valid as for JupyterLab 3.0.0b8

**Comment by jasongrout** [comment]
Can you look into this? With the stock completer and jlab 3.0.0rc1, it seems that things are being sorted according to the match score in my simple example, where I typed `tri<tab>m` - notice that `trim` comes before `tril_indices_from`:
<img width="542" alt="Screen Shot 2020-09-28 at 9 49 22 PM" src="https://user-images.githubusercontent.com/192614/94513874-ad78c600-01d4-11eb-89e0-7f6b913732c9.png">

**Comment by jasongrout** [comment]
Again, with the stock completer in 3.0rc1:

![Uploading Screen Shot 2020-09-28 at 9.52.30 PM.png…]()

**Comment by krassowski** [comment]
No, this is a _different_ completer. You are looking at the result of [`CompleterModel._filter(): IIterator<Completer.IItem>`](https://github.com/jupyterlab/jupyterlab/blob/2e10aae88634e496c1506176fbfaca7eeadf55f6/packages/completer/src/model.ts#L434) which is fine - it does filter *and* sort items.

The new, sadly co-existing (sadly because the change of including this in 2.2 was breaking anyways - we could have replace the "old" way back then) does not sort and is implemented differently; the equivalent of `items()` or `_filter()` is [`_markup(query: string): CompletionHandler.ICompletionItems`](https://github.com/jupyterlab/jupyterlab/blob/2e10aae88634e496c1506176fbfaca7eeadf55f6/packages/completer/src/model.ts#L397) which *does* filter but does not sort.

Both the old `_filter()` method and the new `_markup()` method do call `StringExt.matchSumOfSquares` but while the former uses the sort score, the latter ignores it.

Just for more context, the _old_ method is called from [`items(): IIterator<Completer.IItem>`](https://github.com/jupyterlab/jupyterlab/blob/2e10aae88634e496c1506176fbfaca7eeadf55f6/packages/completer/src/model.ts#L209) while the _new_ one is called from [`completionItems?(): CompletionHandler.ICompletionItems`](https://github.com/jupyterlab/jupyterlab/blob/fa86276692e559525a1e3e8f407e1d17b2adaf97/packages/completer/src/model.ts#L175).

I think that 3.x would be a good point to drop the old one (so that we do not have to support something which is not being used in core, i.e. the new way of providing completions - which is cleaner and more extendible - as this leads to misunderstandings as above), but this is blocked by this sorting issue IMO.

I am sorry I would not be able to PR today nor tomorrow. The fix is potentailly trivial, but I am really swamped. I might be able later but no promises.

**Comment by jasongrout** [comment]
Thanks for the additional context. When I looked at the `_markup` method, I was puzzled that it did not seem to use the score to sort, nor did it check that the score was null to filter (instead checked score as a boolean). Now that you point it out, I see the `_filter` method below it that seems very similar, and I agree that it is odd that we have both `.items` and `.completionItems`.

CC @edzkite, @saulshanabrook

**Comment by jasongrout** [comment]
> I am sorry I would not be able to PR today nor tomorrow. The fix is potentailly trivial, but I am really swamped. I might be able later but no promises.

It looks like the sorting fix can be done in a patch release, so does not *have* to be done before 3.0 this week, though if you get to it, that would be great, of course. I agree it's too bad we didn't drop the old api for 3.0. I won't have time to dive into this either (i.e., to be confident of a fix without unintended consequences), so it sounds like dropping the old api will have to wait until 4.0.

**Comment by jasongrout** [comment]
> The fix is potentailly trivial

I put in a trivial fix in #9098, but I have not tested it. Can you test it before 3.0?

**Comment by edzkite** [comment]
Thanks for bringing this up! For background, when we were implementing the new API, the intent was to enable extensions to control the sort order, as opposed to Jupyter Lab itself. See [here](https://github.com/jupyterlab/jupyterlab/issues/7044#issuecomment-594249225) for more context.

In actuality, we now have a system where the old API filters, marks up, and sorts completions while the new API does not sort. As we were authoring our extension, we also clashed with the filtering and mark up logic, so we ended up monkey patching the model class entirely:
- https://github.com/kiteco/jupyterlab-kite/blob/2b6ea09b7892d751dce9c9b575246c9b2c4bacf8/packages/jupyterlab-kite/src/adapters/jupyterlab/jl_adapter.ts#L275
- https://github.com/kiteco/jupyterlab-kite/blob/2b6ea09b7892d751dce9c9b575246c9b2c4bacf8/packages/jupyterlab-kite/src/adapters/jupyterlab/kite_model.ts

I'm in favor of re-adding the sorting logic back into the completer, but it'd be nice if we had more extensibility into the model by making the sorting and filtering accessible rather than private, so that extension developers can custom them to their needs.
